lib/language_pack/helpers/outdated_ruby_version.rb

```diff
@@ -0,0 +1,150 @@
+# Queries S3 in the background to determine
+# what versions are supported so they can be recommended
+# to the user
+#
+# Example:
+#
+#   ruby_version = LanguagePack::RubyVersion.new("ruby-2.2.5")
+#   outdated = LanguagePack::Helpers::OutdatedRubyVersion.new(
+#     current_ruby_version: ruby_version,
+#     fetcher: LanguagePack::Fetcher.new(LanguagePack::Base::VENDOR_URL, "heroku-16")
+#   )
+#
+#   outdated.call
+#   puts outdated.suggested_ruby_minor_version
+#   #=> "ruby-2.2.10"
+class LanguagePack::Helpers::OutdatedRubyVersion
+  DEFAULT_RANGE = 1..5
+  attr_reader :current_ruby_version
+
+  def initialize(current_ruby_version: , fetcher:)
+    @current_ruby_version = current_ruby_version
+    @fetcher      = fetcher
+
+    @minor_versions = [current_ruby_version]
+    @eol_versions = []
+    @threads = []
+  end
+
+  # Enqueues checks in the background
+  def call
+    return unless can_check?
+
+    check_minor_versions
+    check_eol_versions_major
+    check_eol_versions_minor
+    self
+  end
+
+  def join
+    return false if current_ruby_version.patchlevel_is_significant?
+    return false if current_ruby_version.rbx?
+    return false if current_ruby_version.jruby?
+
+    @threads.each(&:join)
```

Comment:
I didn't realize the threads were sharing memory. I think this might need some changing. Another option better than raw Threads and fork-join is [Concurrent::Future](https://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/Future.html)

Reply:
Moving convo over to https://github.com/heroku/heroku-buildpack-ruby/pull/864#discussion_r329653044